### ISSUE TITLE
SBT 1.3.9 "your proxy requires authentication"

### ISSUE THREAD

**Issue by louiselmps** [sbt/sbt#5496]
## steps
(running on windows 10)
1. Installed java (java version "1.8.0_241")
2. Installed sbt msi latest version (1.3.9)
3. Created a new project folder, and inside that folder i tried to create a new project with "sbt" command, and
it gives the following error:

## problem
``` [info] [launcher] getting org.scala-sbt sbt 1.3.9  (this may take some time)...
:: loading settings :: url = jar:file:/C:/Program%20Files%20(x86)/sbt/bin/sbt-launch.jar!/org/apache/ivy/core/settings/ivysettings.xml
:: loading settings :: url = jar:file:/C:/Program%20Files%20(x86)/sbt/bin/sbt-launch.jar!/org/apache/ivy/core/settings/ivysettings.xml
:: loading settings :: url = jar:file:/C:/Program%20Files%20(x86)/sbt/bin/sbt-launch.jar!/org/apache/ivy/core/settings/ivysettings.xml
:: loading settings :: url = jar:file:/C:/Program%20Files%20(x86)/sbt/bin/sbt-launch.jar!/org/apache/ivy/core/settings/ivysettings.xml
:: loading settings :: url = jar:file:/C:/Program%20Files%20(x86)/sbt/bin/sbt-launch.jar!/org/apache/ivy/core/settings/ivysettings.xml
:: loading settings :: url = jar:file:/C:/Program%20Files%20(x86)/sbt/bin/sbt-launch.jar!/org/apache/ivy/core/settings/ivysettings.xml
:: loading settings :: url = jar:file:/C:/Program%20Files%20(x86)/sbt/bin/sbt-launch.jar!/org/apache/ivy/core/settings/ivysettings.xml
:: loading settings :: url = jar:file:/C:/Program%20Files%20(x86)/sbt/bin/sbt-launch.jar!/org/apache/ivy/core/settings/ivysettings.xml

:: problems summary ::
:::: WARNINGS
        Your proxy requires authentication.

        Your proxy requires authentication.

        Your proxy requires authentication.

        Your proxy requires authentication.

        Your proxy requires authentication.

        Your proxy requires authentication.

                module not found: com.eed3si9n#sjson-new-scalajson_2.12;0.8.2

        ==== local: tried

          C:\Users\lumapas.ml\.ivy2\local\com.eed3si9n\sjson-new-scalajson_2.12\0.8.2\ivys\ivy.xml

          -- artifact com.eed3si9n#sjson-new-scalajson_2.12;0.8.2!sjson-new-scalajson_2.12.jar:

          C:\Users\lumapas.ml\.ivy2\local\com.eed3si9n\sjson-new-scalajson_2.12\0.8.2\jars\sjson-new-scalajson_2.12.jar

        ==== local-preloaded-ivy: tried

          file:/C:/Users/lumapas.ml/.sbt/preloaded/com.eed3si9n/sjson-new-scalajson_2.12/0.8.2/ivys/ivy.xml

        ==== local-preloaded: tried

          file:/C:/Users/lumapas.ml/.sbt/preloaded/com/eed3si9n/sjson-new-scalajson_2.12/0.8.2/sjson-new-scalajson_2.12-0.8.2.pom

          -- artifact com.eed3si9n#sjson-new-scalajson_2.12;0.8.2!sjson-new-scalajson_2.12.jar:

          file:/C:/Users/lumapas.ml/.sbt/preloaded/com/eed3si9n/sjson-new-scalajson_2.12/0.8.2/sjson-new-scalajson_2.12-0.8.2.jar

        ==== Maven Central: tried

          https://repo1.maven.org/maven2/com/eed3si9n/sjson-new-scalajson_2.12/0.8.2/sjson-new-scalajson_2.12-0.8.2.pom

          -- artifact com.eed3si9n#sjson-new-scalajson_2.12;0.8.2!sjson-new-scalajson_2.12.jar:

          https://repo1.maven.org/maven2/com/eed3si9n/sjson-new-scalajson_2.12/0.8.2/sjson-new-scalajson_2.12-0.8.2.jar

        ==== sbt-maven-releases: tried

          https://repo.scala-sbt.org/scalasbt/maven-releases/com/eed3si9n/sjson-new-scalajson_2.12/0.8.2/sjson-new-scalajson_2.12-0.8.2.pom

          -- artifact com.eed3si9n#sjson-new-scalajson_2.12;0.8.2!sjson-new-scalajson_2.12.jar:

          https://repo.scala-sbt.org/scalasbt/maven-releases/com/eed3si9n/sjson-new-scalajson_2.12/0.8.2/sjson-new-scalajson_2.12-0.8.2.jar

        ==== sbt-maven-snapshots: tried

          https://repo.scala-sbt.org/scalasbt/maven-snapshots/com/eed3si9n/sjson-new-scalajson_2.12/0.8.2/sjson-new-scalajson_2.12-0.8.2.pom

          -- artifact com.eed3si9n#sjson-new-scalajson_2.12;0.8.2!sjson-new-scalajson_2.12.jar:

          https://repo.scala-sbt.org/scalasbt/maven-snapshots/com/eed3si9n/sjson-new-scalajson_2.12/0.8.2/sjson-new-scalajson_2.12-0.8.2.jar

        ==== typesafe-ivy-releases: tried

          https://repo.typesafe.com/typesafe/ivy-releases/com.eed3si9n/sjson-new-scalajson_2.12/0.8.2/ivys/ivy.xml

        ==== sbt-ivy-snapshots: tried

          https://repo.scala-sbt.org/scalasbt/ivy-snapshots/com.eed3si9n/sjson-new-scalajson_2.12/0.8.2/ivys/ivy.xml

                ::::::::::::::::::::::::::::::::::::::::::::::

                ::          UNRESOLVED DEPENDENCIES         ::

                ::::::::::::::::::::::::::::::::::::::::::::::

                :: com.eed3si9n#sjson-new-scalajson_2.12;0.8.2: not found

                ::::::::::::::::::::::::::::::::::::::::::::::


:::: ERRORS
        Server access Error: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target url=https://repo1.maven.org/maven2/com/eed3si9n/sjson-new-scalajson_2.12/0.8.2/sjson-new-scalajson_2.12-0.8.2.pom

        Server access Error: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target url=https://repo1.maven.org/maven2/com/eed3si9n/sjson-new-scalajson_2.12/0.8.2/sjson-new-scalajson_2.12-0.8.2.jar


:: USE VERBOSE OR DEBUG MESSAGE LEVEL FOR MORE DETAILS
[error] [launcher] unresolved dependency: com.eed3si9n#sjson-new-scalajson_2.12;0.8.2: not found
[error] [launcher] error during sbt launcher: error retrieving required libraries
  (see C:\Users\lumapas.ml\.sbt\boot\update.log for complete log)
[error] [launcher] could not retrieve sbt 1.3.9
```

## expectation

It should have downloaded all the necessary files and directories

## notes
1. **important:**  Im running through this application under my company's proxy network
2. My project directory has no files yet as i'm trying to build a Scala application from scratch. 

## What i already did but did not work

1. I already added these properties on both sbtconfig and sbtopts files(C:\Program Files (x86)\sbt\conf). (asterisks because of security reasons)

-Dsbt.gigahorse=false
   
-Dhttp.proxyHost=*****
-Dhttp.proxyPort=3128
-Dhttp.proxyUser=*****
-Dhttp.proxyPassword=*****
-Dhttp.proxySet=true

-Dhttps.proxyHost=****
-Dhttps.proxyPort=3128
-Dhttps.proxyUser=******
-Dhttps.proxyPassword=*****
-Dhttps.proxySet=true

2. also tried setting both of these in environment variables, but not simultaneously
JAVA_OPTS= "-Dhttp.proxyHost=*** -Dhttp.proxyPort=3128 -Dhttp.proxyUser=***** -Dhttp.proxyPassword=*****"

SBT_OPTS="-Dhttp.proxyHost=*** -Dhttp.proxyPort=3128 -Dhttp.proxyUser=***** -Dhttp.proxyPassword=*****"

**Comment by eed3si9n** [comment]
@louiselumapas Thanks for the report. sbt version is determined by the content of `project/build.properties` file. Does your project have `project/build.properties`? Does it specify 1.3.9?

**Comment by louiselmps** [comment]
The project has no build.properties file. Like i stated, im starting off on a new project  and currently has no existing files.

**Comment by eed3si9n** [comment]
I am guessing that this is leaking from older Contraband used on IO.
https://github.com/sbt/contraband/blob/v0.4.1/plugin/src/main/scala/ContrabandPlugin.scala#L106
I'll bump up the Contraband version and see if that fixes it.

**Comment by eed3si9n** [comment]
Bumping Contraband didn't work. It looks like this is a problem with using Coursier-built preloaded local from Ivy.

**Comment by ekrich** [comment]
We are trying to get SBT to work behind a Basic User Auth proxy and still having no luck. We are pulling the following container so it is up to date.
https://hub.docker.com/layers/hseeberger/scala-sbt/8u282_1.5.3_2.12.13/images/sha256-3507766194bebe3b725c1e90f13bc833db1cf16967fce4bec1a1cc98c2bd2c82?context=explore

We are exporting JAVA_OPTS with our tried and true -D proxy settings. We got coursier HTTP 407 errors so we tried to use Ivy as shown and that doesn't work either.

I still have problems with Metals - I don't think that coursier, maybe the launcher cannot use JAVA_OPTS properly - it seem to find the first URL but then if it is redirected, the proxy info is not passed on the second URL. This is closed but not fixed 100% - https://github.com/coursier/coursier/issues/930

Shown here as well - https://github.com/scalameta/metals/issues/2485

```scala
+ sbt -Dsbt.coursier=false -Dsbt.ci=true -Djavax.net.ssl.trustStore=./cacerts-maven compile
[info] [launcher] getting org.scala-sbt sbt 1.4.9  (this may take some time)...
[info] [launcher] getting Scala 2.12.12 (for sbt)...
[0m[[0m[0minfo[0m] [0m[0mwelcome to sbt 1.4.9 (Oracle Corporation Java 1.8.0_282)[0m
[0m[[0m[0minfo[0m] [0m[0mloading settings for project prairie_spark-hive-insert_master-build from plugins.sbt ...[0m
[0m[[0m[0minfo[0m] [0m[0mloading project definition from /home/jenkins/agent/workspace/prairie_spark-hive-insert_master/project[0m
[0m[[0m[33mwarn[0m] [0m[0mYour proxy requires authentication.[0m
[0m[[0m[33mwarn[0m] [0m[0mYour proxy requires authentication.[0m
[0m[[0m[33mwarn[0m] [0m[0m	module not found: com.eed3si9n#sbt-assembly;0.15.0[0m
[0m[[0m[33mwarn[0m] [0m[0m==== public: tried[0m
[0m[[0m[33mwarn[0m] [0m[0m  https://repo1.maven.org/maven2/com/eed3si9n/sbt-assembly_2.12_1.0/0.15.0/sbt-assembly-0.15.0.pom[0m
[0m[[0m[33mwarn[0m] [0m[0m==== sbt-plugin-releases: tried[0m
[0m[[0m[33mwarn[0m] [0m[0m  https://repo.scala-sbt.org/scalasbt/sbt-plugin-releases/com.eed3si9n/sbt-assembly/scala_2.12/sbt_1.0/0.15.0/ivys/ivy.xml[0m
[0m[[0m[33mwarn[0m] [0m[0m==== typesafe-alt-ivy-releases: tried[0m
[0m[[0m[33mwarn[0m] [0m[0m  https://repo.typesafe.com/typesafe/ivy-releases/com.eed3si9n/sbt-assembly/scala_2.12/sbt_1.0/0.15.0/ivys/ivy.xml[0m
[0m[[0m[33mwarn[0m] [0m[0m==== local: tried[0m
[0m[[0m[33mwarn[0m] [0m[0m  /root/.ivy2/local/com.eed3si9n/sbt-assembly/scala_2.12/sbt_1.0/0.15.0/ivys/ivy.xml[0m
[0m[[0m[33mwarn[0m] [0m[0m==== local-preloaded-ivy: tried[0m
[0m[[0m[33mwarn[0m] [0m[0m  /root/.sbt/preloaded/com.eed3si9n/sbt-assembly/0.15.0/ivys/ivy.xml[0m
[0m[[0m[33mwarn[0m] [0m[0m==== local-preloaded: tried[0m
[0m[[0m[33mwarn[0m] [0m[0m  file:////root/.sbt/preloaded/com/eed3si9n/sbt-assembly_2.12_1.0/0.15.0/sbt-assembly-0.15.0.pom[0m
[0m[[0m[33mwarn[0m] [0m[0mYour proxy requires authentication.[0m
[0m[[0m[33mwarn[0m] [0m[0mYour proxy requires authentication.[0m
[0m[[0m[33mwarn[0m] [0m[0mYour proxy requires authentication.[0m
[0m[[0m[33mwarn[0m] [0m[0mYour proxy requires authentication.[0m
[0m[[0m[33mwarn[0m] [0m[0mYour proxy requires authentication.[0m
[0m[[0m[33mwarn[0m] [0m[0mYour proxy requires authentication.[0m
[0m[[0m[33mwarn[0m] [0m[0mYour proxy requires authentication.[0m
[0m[[0m[33mwarn[0m] [0m[0mYour proxy requires authentication.[0m
[0m[[0m[33mwarn[0m] [0m[0m	::::::::::::::::::::::::::::::::::::::::::::::[0m
[0m[[0m[33mwarn[0m] [0m[0m	::          UNRESOLVED DEPENDENCIES         ::[0m
[0m[[0m[33mwarn[0m] [0m[0m	::::::::::::::::::::::::::::::::::::::::::::::[0m
[0m[[0m[33mwarn[0m] [0m[0m	:: com.eed3si9n#sbt-assembly;0.15.0: not found[0m
[0m[[0m[33mwarn[0m] [0m[0m	::::::::::::::::::::::::::::::::::::::::::::::[0m
[0m[[0m[33mwarn[0m] [0m[0m[0m
[0m[[0m[33mwarn[0m] [0m[0m	Note: Some unresolved dependencies have extra attributes.  Check that these dependencies exist with the requested attributes.[0m
[0m[[0m[33mwarn[0m] [0m[0m		com.eed3si9n:sbt-assembly:0.15.0 (scalaVersion=2.12, sbtVersion=1.0)[0m
[0m[[0m[33mwarn[0m] [0m[0m[0m
[0m[[0m[33mwarn[0m] [0m[0m	Note: Unresolved dependencies path:[0m
[0m[[0m[33mwarn[0m] [0m[0m		com.eed3si9n:sbt-assembly:0.15.0 (scalaVersion=2.12, sbtVersion=1.0) (/home/jenkins/agent/workspace/prairie_spark-hive-insert_master/project/plugins.sbt#L1-2)[0m
[0m[[0m[33mwarn[0m] [0m[0m		  +- default:prairie_spark-hive-insert_master-build:0.1.0-SNAPSHOT (scalaVersion=2.12, sbtVersion=1.0)[0m
[0m[[0m[31merror[0m] [0m[0msbt.librarymanagement.ResolveException: unresolved dependency: com.eed3si9n#sbt-assembly;0.15.0: not found[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.librarymanagement.IvyActions$.resolveAndRetrieve(IvyActions.scala:335)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.librarymanagement.IvyActions$.$anonfun$updateEither$1(IvyActions.scala:211)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.librarymanagement.IvySbt$Module.$anonfun$withModule$1(Ivy.scala:251)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.librarymanagement.IvySbt.$anonfun$withIvy$1(Ivy.scala:215)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.librarymanagement.IvySbt.sbt$internal$librarymanagement$IvySbt$$action$1(Ivy.scala:77)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.librarymanagement.IvySbt$$anon$1.call(Ivy.scala:87)[0m
[0m[[0m[31merror[0m] [0m[0m	at xsbt.boot.Locks$GlobalLock.withChannel$1(Locks.scala:113)[0m
[0m[[0m[31merror[0m] [0m[0m	at xsbt.boot.Locks$GlobalLock.withChannelRetries$1(Locks.scala:91)[0m
[0m[[0m[31merror[0m] [0m[0m	at xsbt.boot.Locks$GlobalLock.$anonfun$withFileLock$1(Locks.scala:119)[0m
[0m[[0m[31merror[0m] [0m[0m	at xsbt.boot.Using$.withResource(Using.scala:12)[0m
[0m[[0m[31merror[0m] [0m[0m	at xsbt.boot.Using$.apply(Using.scala:9)[0m
[0m[[0m[31merror[0m] [0m[0m	at xsbt.boot.Locks$GlobalLock.withFileLock(Locks.scala:119)[0m
[0m[[0m[31merror[0m] [0m[0m	at xsbt.boot.Locks$GlobalLock.ignoringDeadlockAvoided(Locks.scala:71)[0m
[0m[[0m[31merror[0m] [0m[0m	at xsbt.boot.Locks$GlobalLock.withLock(Locks.scala:59)[0m
[0m[[0m[31merror[0m] [0m[0m	at xsbt.boot.Locks$.apply0(Locks.scala:47)[0m
[0m[[0m[31merror[0m] [0m[0m	at xsbt.boot.Locks$.apply(Locks.scala:36)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.librarymanagement.IvySbt.withDefaultLogger(Ivy.scala:87)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.librarymanagement.IvySbt.withIvy(Ivy.scala:209)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.librarymanagement.IvySbt.withIvy(Ivy.scala:206)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.librarymanagement.IvySbt$Module.withModule(Ivy.scala:250)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.librarymanagement.IvyActions$.updateEither(IvyActions.scala:196)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.librarymanagement.ivy.IvyDependencyResolution.update(IvyDependencyResolution.scala:23)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.librarymanagement.DependencyResolution.update(DependencyResolution.scala:60)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.LibraryManagement$.resolve$1(LibraryManagement.scala:53)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$12(LibraryManagement.scala:103)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.util.Tracked$.$anonfun$lastOutput$1(Tracked.scala:73)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$20(LibraryManagement.scala:116)[0m
[0m[[0m[31merror[0m] [0m[0m	at scala.util.control.Exception$Catch.apply(Exception.scala:228)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$11(LibraryManagement.scala:116)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$11$adapted(LibraryManagement.scala:97)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.util.Tracked$.$anonfun$inputChangedW$1(Tracked.scala:219)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.LibraryManagement$.cachedUpdate(LibraryManagement.scala:130)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.Classpaths$.$anonfun$updateTask0$5(Defaults.scala:3525)[0m
[0m[[0m[31merror[0m] [0m[0m	at scala.Function1.$anonfun$compose$1(Function1.scala:49)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.std.Transform$$anon$4.work(Transform.scala:68)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.Execute.$anonfun$submit$2(Execute.scala:282)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.Execute.work(Execute.scala:291)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.Execute.$anonfun$submit$1(Execute.scala:282)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)[0m
[0m[[0m[31merror[0m] [0m[0m	at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)[0m
[0m[[0m[31merror[0m] [0m[0m	at java.util.concurrent.FutureTask.run(FutureTask.java:266)[0m
[0m[[0m[31merror[0m] [0m[0m	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[0m
[0m[[0m[31merror[0m] [0m[0m	at java.util.concurrent.FutureTask.run(FutureTask.java:266)[0m
[0m[[0m[31merror[0m] [0m[0m	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)[0m
[0m[[0m[31merror[0m] [0m[0m	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)[0m
[0m[[0m[31merror[0m] [0m[0m	at java.lang.Thread.run(Thread.java:748)[0m
[0m[[0m[31merror[0m] [0m[0m([31mupdate[0m) sbt.librarymanagement.ResolveException: unresolved dependency: com.eed3si9n#sbt-assembly;0.15.0: not found[0m
[0m[[0m[33mwarn[0m] [0m[0mProject loading failed: (r)etry, (q)uit, (l)ast, or (i)gnore? (default: r)[0m
```

**Comment by ekrich** [comment]
I can tell you definitely that even if a direct call by coursier will work with a basic auth proxy, but if that call results in a redirect, the redirect fails. The header containing the basic auth needs to be passed on all redirects and forwards.

See https://github.com/coursier/coursier/issues/930

At the time I reported this where the redirect fails.

> You can try in a browser too.
If you go to https://repo.scala-sbt.org/scalasbt
it should redirect to https://dl.bintray.com/sbt/

So basically, things like Metals/Bloop won't install on a corporate network using basic auth because of this redirect problem. 

Therefore it is astronomically harder to get Scala in the door for any corporation with this issue. If someone knows how to work on coursier, it is probably an easy fix.

**Comment by ekrich** [comment]
This was the log the last time I tested.
[metals.txt](https://github.com/sbt/sbt/files/10559780/metals.txt)

**Comment by trevorprater** [comment]
Is there a solution here? I am experiencing this issue, too. I’m not sure how to proceed or work around this, unfortunately.

**Comment by ekrich** [comment]
The only workaround I know is that if you can use your computer off the corporate network, then you can get it working and then it should work until there is a new Metals update.

If you don't have that option, you may be able to use the Ivy dependency manager which works but there is talk of that being removed in sbt 2.0. I honestly don't think that will work for Metals but may work if you have problems in your build. Temporarily add the following to your sbt file.

`ThisBuild / useCoursier := false`
